VoodooPS2Controller/VoodooPS2Controller.cpp

```diff
@@ -1088,74 +1116,53 @@ void ApplePS2Controller::processRequest(PS2Request * request)
-        request->commands[index].inOrOut = readDataPort(deviceMode);
+        request->commands[index].inOrOut = readDataPort(devicePort);
         break;
 
       case kPS2C_ReadDataPortAndCompare:
 #if OUT_OF_ORDER_DATA_CORRECTION_FEATURE
-        byte = readDataPort(deviceMode, request->commands[index].inOrOut);
+        byte = readDataPort(devicePort, request->commands[index].inOrOut);
 #else 
-        byte = readDataPort(deviceMode);
+        byte = readDataPort(devicePort);
 #endif
         failed = (byte != request->commands[index].inOrOut);
         request->commands[index].inOrOut = byte;
         break;
 
       case kPS2C_WriteDataPort:
-        writeDataPort(request->commands[index].inOrOut);
-        if (transmitToMouse)     // next reads from mouse input stream
-        {
-          deviceMode      = kDT_Mouse;
-          transmitToMouse = false;
+        if (devicePort == kPS2AuxIdx) {
+          writeCommandPort(kCP_TransmitToMouse);
+        } else if (devicePort > kPS2AuxIdx) {
+          writeCommandPort(kCP_TransmitToMuxedMouse + (devicePort - kPS2MuxIdx));
         }
-        else
-        {
-           deviceMode   = kDT_Keyboard;
-        }
-        break;
-
-      case kPS2C_WriteCommandPort:
-        writeCommandPort(request->commands[index].inOrOut);
-        if (request->commands[index].inOrOut == kCP_TransmitToMouse)
-          transmitToMouse = true; // preparing to transmit data to mouse
+        
+        writeDataPort(request->commands[index].inOrOut);
         break;
 
       //
       // Send a composite mouse command that is equivalent to the following
       // (frequently used) command sequence:
       //
-      // 1. kPS2C_WriteCommandPort( kCP_TransmitToMouse )
-      // 2. kPS2C_WriteDataPort( command )
-      // 3. kPS2C_ReadDataPortAndCompare( kSC_Acknowledge )
+      // 1. kPS2C_WriteDataPort( command )
+      // 2. kPS2C_ReadDataPortAndCompare( kSC_Acknowledge )
       //
 
-      case kPS2C_SendMouseCommandAndCompareAck:
-        writeCommandPort(kCP_TransmitToMouse);
+      case kPS2C_SendCommandAndCompareAck:
+        if (devicePort == kPS2AuxIdx) {
+          writeCommandPort(kCP_TransmitToMouse);
+        } else if (devicePort > kPS2AuxIdx) {
+          writeCommandPort(kCP_TransmitToMuxedMouse + (devicePort - kPS2MuxIdx));
+        }
+        
         writeDataPort(request->commands[index].inOrOut);
-        deviceMode = kDT_Mouse;
 #if OUT_OF_ORDER_DATA_CORRECTION_FEATURE
-        byte = readDataPort(kDT_Mouse, kSC_Acknowledge);
+        byte = readDataPort(devicePort, kSC_Acknowledge);
 #else 
-        byte = readDataPort(kDT_Mouse);
+        byte = readDataPort(devicePort);
 #endif
         failed = (byte != kSC_Acknowledge);
         break;
             
-      case kPS2C_ReadMouseDataPort:
-        deviceMode= kDT_Mouse;
-        request->commands[index].inOrOut = readDataPort(deviceMode);
-        break;
-            
-      case kPS2C_ReadMouseDataPortAndCompare:
```

Comment:
Why do we no longer need these commands even for setups without mux?

Reply:
The only difference between talking to a mouse compared to the keyboard is that an additional command (kCP_TransmitToMouse) is written to the command port. Talking to a muxed device also requires an additional command (kCP_TransmitToMuxedMouse). Otherwise the logic is the same between the keyboard and mouse.

It made more sense to me to write to the command port based off of the port number of the request. This made it so that the trackpad drivers didn't care about the difference between a mouse and keyboard command. There shouldn't be any reason that the trackpad should be talking to the keyboard port anyways.